usr.bin/monitor/monitor.c

```diff
@@ -0,0 +1,248 @@
+ *    distribution.
+ * 3. Neither the name of The DragonFly Project nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific, prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ */
+
+#include <sys/types.h>
+#include <sys/event.h>
+#include <sys/time.h>
+#include <sys/stat.h>
+#include <stdio.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <fcntl.h>
+
+typedef struct monitor_elm {
+	const char *path;
+	int	fd;
+} *monitor_elm_t;
+
+static void usage(int exit_code);
+static void monitor_add(const char *path);
+static void monitor_events(void);
+
+static int VerboseOpt;
+static int QuietOpt;
+static int ExitOpt;
+static int KQueueFd;
+static int NumFiles;
+static int MaxFiles;
+static monitor_elm_t *Elms;
+
+int
+main(int ac, char **av)
+{
+	int ch;
+	int i;
+
+	while ((ch = getopt(ac, av, "qvx")) != -1) {
+		switch (ch) {
+		case 'q':
+			if (VerboseOpt > 0)
+				--VerboseOpt;
+			else
+				++QuietOpt;
+			break;
+		case 'v':
+			if (QuietOpt > 0)
+				--QuietOpt;
+			else
+				++VerboseOpt;
+			break;
+		case 'x':
+			ExitOpt = 1;
+			break;
+		default:
+			usage(1);
+			/* not reached */
+		}
+	}
+	ac -= optind;
+	av += optind;
+
+	if (ac < 1) {
+		usage(1);
+		/* not reached */
+	}
+
+	if ((KQueueFd = kqueue()) < 0) {
+		perror("kqueue");
+		exit(1);
+	}
+	NumFiles = MaxFiles = 16;
+	Elms = calloc(MaxFiles, sizeof(monitor_elm_t));
+
+	for (i = 0; i < ac; ++i) {
+		monitor_add(av[i]);
+	}
+	fflush(stdout);
+	do {
+		monitor_events();
+		fflush(stdout);
+	} while (ExitOpt == 0);
+	exit(0);
+}
+
+static
+void
+monitor_add(const char *path)
+{
+	monitor_elm_t elm;
+	struct kevent kev;
+	int n;
+
+	elm = malloc(sizeof(*elm));
+	bzero(elm, sizeof(*elm));
+	elm->path = path;
+	elm->fd = open(path, O_RDONLY);
+	if (elm->fd < 0) {
+		printf("%s\tnot found\n", path);
+		return;
+	}
+	EV_SET(&kev, elm->fd, EVFILT_VNODE, EV_ADD | EV_ENABLE | EV_CLEAR,
+		NOTE_DELETE | NOTE_WRITE | NOTE_EXTEND | NOTE_ATTRIB |
+		NOTE_LINK | NOTE_RENAME | NOTE_REVOKE |
+#if defined(__FreeBSD__) || defined(__NetBSD__)
```

Comment:
It would be better to `#ifdef` on the flag being tested rather than on the operating system.

Reply:
> It would be better to `#ifdef` on the flag being tested rather than on the operating system.

In that case I think an `#ifdef NOTE_OPEN` should be enough. WDYT?

Reply:
> It would be better to `#ifdef` on the flag being tested rather than on the operating system.

Removed the ifdef's as they pollute the code.